cmd/skopeo/sync.go

```diff
@@ -0,0 +1,509 @@
+	if err != nil {
+		return fmt.Errorf("Error loading trust policy: %v", err)
+	}
+	defer policyContext.Destroy()
+
+	destRef, err := getImageReference(c.Args()[0])
+	if err != nil {
+		return fmt.Errorf("Error while parsing destination: %v", err)
+	}
+	destinationCtx, err := contextFromGlobalOptions(c, "dest-")
+	if err != nil {
+		return err
+	}
+
+	signBy := c.String("sign-by")
+	removeSignatures := c.Bool("remove-signatures")
+
+	//TODO: should we assume that's our default manifest type?
+	manifestType := manifest.DockerV2Schema2MediaType
+
+	var srcRepoList []repoDescriptor
+
+	if c.IsSet("source") {
+		srcRepo, err := syncSourceHandler(c, destRef)
+		if err != nil {
+			return err
+		}
+		srcRepoList = append(srcRepoList, srcRepo)
+	} else {
+		srcRepoList, err = syncSourceFileHandler(c)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, srcRepo := range srcRepoList {
+		options := copy.Options{
+			RemoveSignatures:      removeSignatures,
+			SignBy:                signBy,
+			ReportWriter:          os.Stdout,
+			DestinationCtx:        destinationCtx,
+			ForceManifestMIMEType: manifestType,
+			SourceCtx:             srcRepo.Context,
+		}
+
+		for counter, ref := range srcRepo.TagsReferences {
+			logrus.WithFields(logrus.Fields{
+				"repo": ref.DockerReference().Name(),
+			}).Infof("Processing tag %d/%d", counter+1, len(srcRepo.TagsReferences))
+			destRef, err := buildFinalDestination(srcRepo.RootReference, ref, c.Args()[0])
+			if err != nil {
+				return err
+			}
+			logrus.WithFields(logrus.Fields{
+				"source":      ref,
+				"destination": destRef,
+			}).Info("Copy started")
+
+			err = copy.Image(context.Background(), policyContext, destRef, ref, &options)
+			if err != nil {
+				logrus.WithFields(logrus.Fields{
+					"repo": ref.DockerReference().Name(),
+				}).Error("Error copying tag, skipping")
+				logrus.Debug(err)
+				continue
+			}
+		}
+	}
+
+	return nil
+}
+
+var syncCmd = cli.Command{
+	Name:  "sync",
+	Usage: "Sync one or more images from one location to another",
+	Description: fmt.Sprint(`
+
+	Copy all the images from SOURCE to DESTINATION.
+
+	Useful to keep in sync a local docker registry mirror. Can be used
+	to populate also registries running inside of air-gapped environments.
+
+	SOURCE can be either a repository hosted on a docker registry
+	(eg: docker://docker.io/busybox) or a local directory
+	(eg: dir:///media/usb/).
+
+	SOURCE-FILE is a YAML file with a set of source images from different
+	docker registry. Local directory are not supported.
+
+	Skopeo sync will copy all the tags of an image when SOURCE uses the
+	'docker://' transport and no tag is specified.
+
+	DESTINATION can be either a docker registry
+	(eg: docker://my-registry.local.lan) or a local directory
+	(eg: dir:///media/usb).
+
+	When DESTINATION is a local directory one directory per 'image:tag' is going
+	to be created.
+	`),
+	ArgsUsage: "[--source SOURCE] [--source-file SOURCE-FILE] DESTINATION",
```

Comment:
I see why we have two different options, but it looks a bit unnatural to add a mandatory argument in this way.

Could it be like "[--source-yaml] SOURCE DESTINATION"

if `--source-yaml` is specified then `SOURCE` points to a YAML file.

Not ideal either, but in the simplest case you have `skopeo sync docker://... docker://...`.